### PR TITLE
CNAM-435: Documentation added for extractors

### PR DESCRIPTION
**Pull request by Anfor**
I add documentation to explain how extractors are built and how mcoExtractors works.

### CODE REVIEW

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala [hunk 1]
```diff
@@ -15,13 +15,34 @@ trait McoExtractor[EventType <: AnyEvent] extends Extractor[EventType] with McoS
 
   val eventBuilder: EventBuilder
 
+  /** It allows get all need columns from source.
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala]: I think all this documentation should be put on the Extractor trait.

File: Extractors.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
+
+
+# Extractors
+
+Extractors are a kind of jobs that allows to extract the required columns from the sources.
```
Comment (ysebiat) [inline Extractors.md]: I think you need to add a sentence after this one, I would suggest the following: 

_and maps them to the Event it is extracting.

 An Extractor is defined by Event type it is extracting and the Source where the information needed to build the extract Event is._

File: Extractors.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+Each manages different parts of extraction. They are all necessary for the creation of an element of the next level.
+2) Intermediate elements are traits that bring together the basic methods to create a common trait to extract the data from the sources (mco,dcir,ssr,had).In our example, 
+[McoExtractor](https://github.com/X-DataInitiative/SCALPEL-Extraction/blob/master/src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala) 
+is the base for retrieving the necessary data from mco source. These elements inherit the base elements and accept as a parameter a trait of the type EventType
+3) The upper elements  are objects and  they are the entry point of the job itself, they inherit from the middle elements and specialize it in one type of `Event`. 
+ 
+Each basic component has a function:
+ - [Extractor](https://github.com/X-DataInitiative/SCALPEL-Extraction/blob/master/src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala)
+ is a trait that works by filtering, extracting and building.It's main method  `extract ` allows to filter the sources and build a dataset of type `Event`.
```
Comment (ysebiat) [inline Extractors.md]: Its instead of it's.

**Review by ysebiat — CHANGES_REQUESTED** [review]
Some comments.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala (hunk 1 above)
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala]: Yes,  It'll  be better placed.

File: Extractors.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+```
+```scala  
+ override def extractGroupId(r: Row): String = {
+      r.getAs[String](ColNames.EtaNum) + "_" +
+      r.getAs[String](ColNames.RsaNum) + "_" +
+      r.getAs[Int](ColNames.Year).toString
+  }
+```
+
+The above elements are responsible for defining the type of `Event` and must implement at least two values and modify them according to their specificity.
```
Comment (dsun0720) [inline Extractors.md]: The values or methods which should be implemented base on their parent trait. I think your words("at least two values") here is a bit confused.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala [hunk 5]
```diff
@@ -9,14 +9,42 @@ import fr.polytechnique.cmap.cnam.etl.sources.Sources
...
   def builder(row: Row): Seq[Event[EventType]]
 
+  /** It allows get all need columns from source.
+   *
+   * @param sources Source object [[Sources]] that contains all sources.
+   * @return A dataframe with mco columns.
+   */
   def getInput(sources: Sources): DataFrame
 
+  /** It extract columns required to build the event object.
```
Comment (dsun0720) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala]: "it extracts"

**Review by dsun0720 — CHANGES_REQUESTED** [review]

File: Extractors.md (hunk 4 above)
Comment (Anfor) [reply Extractors.md]: ok, this approach is better understood?
"The above elements are responsible for defining the type of `Event` and must implement at least  `columnName` and  `eventBuilder` values and modify them according to their specificity."

Comment (dsun0720) [reply Extractors.md]: yes

**Review by dsun0720 — DISMISSED** [review]

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala [hunk 6]
```diff
@@ -9,14 +9,42 @@ import fr.polytechnique.cmap.cnam.etl.sources.Sources
 
 trait Extractor[EventType <: AnyEvent] extends Serializable {
 
+  /** It allows to check if the passed row is in the codes set.
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala]: I think it is better to phrase it as : _Allows to check if the Row from the Source is considered in the current Study._

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala [hunk 7]
```diff
@@ -9,14 +9,42 @@ import fr.polytechnique.cmap.cnam.etl.sources.Sources
...
+   *
+   * @param codes A set of codes to study.
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala]: _A set of codes being considered in the Study._

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala [hunk 8]
```diff
@@ -9,14 +9,42 @@ import fr.polytechnique.cmap.cnam.etl.sources.Sources
...
+   * @param row   The row itself.
+   * @return A boolean value.
+   */
   def isInStudy(codes: Set[String])(row: Row): Boolean
 
+  /** It checks if the row is null at columnName.
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala]: _Checks if the passed Row has the information needed to build the Event._

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala [hunk 9]
```diff
@@ -9,14 +9,42 @@ import fr.polytechnique.cmap.cnam.etl.sources.Sources
...
+   *
+   * @param row The row itself.
+   * @return A boolean value.
+   */
   def isInExtractorScope(row: Row): Boolean
 
+  /** It build an Event from a row.
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala]: _Builds the Event._

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala [hunk 10]
```diff
@@ -9,14 +9,42 @@ import fr.polytechnique.cmap.cnam.etl.sources.Sources
...
+   *
+   * @param row The row itself.
+   * @return An event object.
+   */
   def builder(row: Row): Seq[Event[EventType]]
 
+  /** It allows get all need columns from source.
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala]: _Gets and prepares all the needed columns from the Source._

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala [hunk 11]
```diff
@@ -9,14 +9,42 @@ import fr.polytechnique.cmap.cnam.etl.sources.Sources
...
+   *
+   * @param sources Source object [[Sources]] that contains all sources.
+   * @return A dataframe with mco columns.
+   */
   def getInput(sources: Sources): DataFrame
 
+  /** It extracts columns required to build the event object.
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/Extractor.scala]: Extracts the Event from the Source.

This function is responsible for glueing different other parts of the Extractor. This method should be considered the unique callable method from a Study perspective.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala [hunk 12]
```diff
@@ -32,19 +32,43 @@ trait McoExtractor[EventType <: AnyEvent] extends Extractor[EventType] with McoS
     Seq(eventBuilder[EventType](patientId, groupId, code(row), weight, eventDate, endDate))
   }
 
+  /** It gets columnName's value from row.
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala]: Extracts the tracked value.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala [hunk 13]
```diff
@@ -32,19 +32,43 @@ trait McoExtractor[EventType <: AnyEvent] extends Extractor[EventType] with McoS
...
+  /** It gets PatientID value from row.
+   *
+   * @param r The row itself.
+   * @return The value of PatientID.
+   */
   def extractPatientId(r: Row): String = {
     r.getAs[String](ColNames.PatientID)
   }
 
+  /** It gets EtaNum value, RsaNum value and Year from row.
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala]: This is wrong description. This is specific to Mco Source and how we group Events from this source together. Better phrasing : 

_Creates an ID that group Events of different categories by concatinating ETA_NUM, RSA_NUM and the YEAR._

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala [hunk 14]
```diff
@@ -32,19 +32,43 @@ trait McoExtractor[EventType <: AnyEvent] extends Extractor[EventType] with McoS
...
+   * @param r The row itself.
+   * @return The value of groupId.
+   */
   override def extractGroupId(r: Row): String = {
     r.getAs[String](ColNames.EtaNum) + "_" +
       r.getAs[String](ColNames.RsaNum) + "_" +
       r.getAs[Int](ColNames.Year).toString
   }
 
+  /** It gets EstimatedStayStart from row.
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/mco/McoExtractor.scala]: _Extracts the EstimatedStayStart as the start._

You will need to document the EstimatedStayStart method, as this an artefact of our code and does not exist as is in the MCO Source.

**Review by ysebiat — CHANGES_REQUESTED** [review]
Changes needed in the phrasing.

**Review by ysebiat — APPROVED** [review]
LGTM.

**Pull request merged**